### DISCUSSION TITLE
How to make Tun2socks work on Debian?

### DISCUSSION THREAD

**Discussion by Tutez64** [xjasonlyu/tun2socks#141] [Q&A]
Hello, I'm having trouble getting Tun2socks to work on my fresh Debian 11.3 VM.

Here is my original ip route

![VirtualBoxVM_6NgZssWdDp](https://user-images.githubusercontent.com/40702709/175440017-19129542-3668-49da-ac14-4a9d457e4867.jpg)

I started by trying this:

```
ip tuntap add mode tun dev tun0
ip addr add 198.18.0.1/15 dev tun0
ip link set dev tun0 up

ip route del default
ip route add default via 198.18.0.1 dev tun0 metric 1
ip route add default via 10.0.2.2 dev enp0s3 metric 10

tun2socks -device tun0 -proxy socks5://host:port -interface enp0s3
```

and then Tun2socks seemed to work but I wasn't able to access websites, Firefox was loading for a long time and then giving me an error:
![VirtualBoxVM_1z2xsFrUmD](https://user-images.githubusercontent.com/40702709/175440610-59af7c4f-cde1-4243-a087-9f0540ab961d.jpg)
![VirtualBoxVM_9uwJavtBRY](https://user-images.githubusercontent.com/40702709/175505134-d3f30b88-4c02-4f0c-9ad0-d862858a31e4.jpg)
![682XoUGbM1](https://user-images.githubusercontent.com/40702709/175443105-160c41fa-7209-4eb4-87ae-c27afa40198d.jpg)
I can ping websites.

I tried adding this:

```
iptables -t nat -A PREROUTING -p udp --dport 53 -j DNAT --to 8.8.8.8:53;
iptables -t nat -A POSTROUTING -p udp -d 8.8.8.8 --dport 53 -o eth3 -j MASQUERADE;
sysctl net.ipv4.conf.all.rp_filter=0
sysctl net.ipv4.conf.enp0s3.rp_filter=0
```

but I saw no change

Then I rebooted and tried this:

```
ip tuntap add mode tun dev tun0
ip addr add 198.18.0.1/15 dev tun0
ip link set dev tun0 up
 
ip route replace default dev tun0
 
tun2socks -device tun0 -proxy socks5://host:port -interface enp0s3
```

![4MTiqxO9R4](https://user-images.githubusercontent.com/40702709/175446626-7cecfae1-9936-4990-a925-84dc2135bb1f.jpg)
![VirtualBoxVM_QOaSqZEMMW](https://user-images.githubusercontent.com/40702709/175447038-c6eaf519-2f83-4416-a4c6-9d99b86725ce.jpg)
Other than that the result is the same, I can ping websites and Firefox give the same error.

I tried adding this again:

```
iptables -t nat -A PREROUTING -p udp --dport 53 -j DNAT --to 8.8.8.8:53;
iptables -t nat -A POSTROUTING -p udp -d 8.8.8.8 --dport 53 -o eth3 -j MASQUERADE;
sysctl net.ipv4.conf.all.rp_filter=0
sysctl net.ipv4.conf.enp0s3.rp_filter=0
```

but I saw no change again

I tried others things like changing my DNS and changing things here and there but none of them worked.

I'm not good at networking so I don't know what am I wrong, I've been trying to find a working proxifier for months (tried Redsocks, Redsocks 2, Tsocks, Proxychains and probably other I forgot) but I either can't make them to work or my IP leaks. Your program seems to be the best option since other are old and often not maintened since years, but I just can't find how to make it works.

**Comment by xjasonlyu** [comment]
Hi, your first trying seems good to me. 

Can you put your route table here after running those commands?

```
ip tuntap add mode tun dev tun0
ip addr add 198.18.0.1/15 dev tun0
ip link set dev tun0 up

ip route del default
ip route add default via 198.18.0.1 dev tun0 metric 1
ip route add default via 10.0.2.2 dev enp0s3 metric 10

tun2socks -device tun0 -proxy socks5://host:port -interface enp0s3
```

it’s weird it shows no route error.

**Reply by Tutez64** [reply]
Oops sorry I put the wrong image, I corrected the message, check again.

**Reply by xjasonlyu** [reply]
Normally, iptables is not needed in this case. Your route table seems fine to me and the log confirms that.

In my opinion, it might be a DNS issue which caused by your socks5 server behind the VM NAT.

Some socks5 server implementation doesn't work properly in those NAT-involved case (for UDP).

Would you mind telling me what the socks5 server you are using?

**Reply by Tutez64** [reply]
I'm using this provider: (edited), I can ask them for info if needed.
